File: C++/0054-spiral-matrix.cpp
class Solution {
 public:
  vector<int> spiralOrder(vector<vector<int>>& matrix) {
    vector<int> list;
    int rows = matrix.size();
    int cols = matrix[0].size();
    int top = 0;
    int bottom = rows - 1;
    int left = 0;
    int right = cols - 1;
    while (list.size() < rows * cols) {
      for (int i = left; i <= right; i++) {
        list.push_back(matrix[top][i]);
      }
      for (int i = top + 1; i <= bottom; i++) {
        list.push_back(matrix[i][right]);
      }
      if (top != bottom) {
        for (int i = right - 1; i >= left; i--) {
          list.push_back(matrix[bottom][i]);
        }
      }
      if (left != right) {
        for (int i = bottom - 1; i > top; i--) {
          list.push_back(matrix[i][left]);
        }
      }
      left++;
      top++;
      right--;
      bottom--;
    }

    return list;
  }
};

File: Python/0054-spiral-matrix.py
# time complexity: O(m*n)
# space complexity: O(1)
from typing import List


class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        ROW = len(matrix)
        COL = len(matrix[0])
        direction = 1
        row = 0
        col = -1
        result = []
        while ROW > 0 and COL > 0:
            for _ in range(COL):
                col += direction
                result.append(matrix[row][col])
            ROW -= 1
            for _ in range(ROW):
                row += direction
                result.append(matrix[row][col])
            COL -= 1

            direction *= -1

        return result


matrix = [[1, 2, 3], [4, 5, 6], [7, 8, 9]]
print(Solution().spiralOrder(matrix))
matrix = [[1, 2, 3, 4], [5, 6, 7, 8], [9, 10, 11, 12]]
print(Solution().spiralOrder(matrix))
